Port udp_driver Changes to serial_driver

JWhitleyWork opened the pull request:
I apologize for the rather large PR. This ports the changes that @reza-ebrahimi, @flynneva, and I have made to `udp_driver` to `serial_driver`. There are some final clean-ups that I would like to do to make sure that the API formatting is consistent between the two and also rename the `io_context` package (it does more than just house the `io_context` class) but this is the bulk of the porting done.

Unfortunately, since serial ports aren't just lying around on the CI systems like network sockets are, it's difficult to add tests which actually open or use a port.

(would appreciate eyes from all/any mentioned here - could only assign @esteve because he is a maintainer on the repo)

esteve commented on io_context/include/msg_converters/example_interfaces.hpp:
```diff
@@ -0,0 +1,36 @@
+namespace common
+{
+
+void from_msg(const UInt8MultiArray::SharedPtr & in, MutBuffer & out);
```
Does the API support messages that are not wrapped in a shared pointer?

JWhitleyWork replied:
It probably makes more sense to pass in a de-referneced pointer from a `::SharedPtr` and not have to do anything for the non-`::SharedPtr` version than to have to wrap the non-`::SharedPtr` version in one. Done.

esteve commented on serial_driver/include/serial_driver/serial_bridge_node.hpp:
```diff
@@ -0,0 +1,98 @@
+
+namespace lc = rclcpp_lifecycle;
+using LNI = rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface;
+using example_interfaces::msg::UInt8MultiArray;
```
Shouldn't this use `serial_msgs`? https://github.com/flynneva/serial_msgs

JWhitleyWork replied:
I looked around for something similar but `UInt8MultiArray` was the closest I could find. We can always switch to it later. I expect that most people won't use the bridge node anyway and will, instead, use the `SerialDriver` directly. Leaving for now.

flynneva commented:
so when I try and run the exe after building this branch i get this error:

```
 ros2 run serial_driver serial_bridge_node_exe 
terminate called after throwing an instance of 'class_loader::LibraryLoadException'
  what():  Could not load library LoadLibrary error: /home/flynn/code/ros/foxy/install/serial_driver/lib/libserial_driver_nodes.so: undefined symbol: _ZN7drivers13serial_driver16SerialBridgeNode11on_shutdownERKN16rclcpp_lifecycle5StateE, at /tmp/binarydeb/ros-foxy-rcutils-1.1.2/src/shared_library.c:84
```

JWhitleyWork commented:
> so when I try and run the exe after building this branch i get this error:
> 
> ```
>  ros2 run serial_driver serial_bridge_node_exe 
> terminate called after throwing an instance of 'class_loader::LibraryLoadException'
>   what():  Could not load library LoadLibrary error: /home/flynn/code/ros/foxy/install/serial_driver/lib/libserial_driver_nodes.so: undefined symbol: _ZN7drivers13serial_driver16SerialBridgeNode11on_shutdownERKN16rclcpp_lifecycle5StateE, at /tmp/binarydeb/ros-foxy-rcutils-1.1.2/src/shared_library.c:84
> ```
> 
> im guessing there probably needs to be some params passed in as well, might be useful to make a `config` directory and store some default yaml file in there

That'll teach me to push without fully testing. I declared the `on_shutdown` function but forgot to define it. That's been fixed now but I will also add an example file in `params` and install it. Maybe even a launch file!

JWhitleyWork commented:
@flynneva Now ready for re-testing.

flynneva commented:
ignore my comments above if you already addressed them (i started a review but never submitted it).

re-compiled and tried to run it and connect to a serial device (a BNO055 IMU) and got this output:
```
ros2 run serial_driver serial_bridge_node_exe --ros-args --params-file ./example.params.yaml 
[INFO] [1615907120.893310599] [IoContext::IoContext]: Thread(s) Created: 2
```

should we add some prints here to help the user understand at what step the driver is at?

flynneva commented on serial_driver/CMakeLists.txt:
```diff
@@ -32,25 +32,55 @@ ament_auto_find_build_dependencies()
+
+rclcpp_components_register_node(${PROJECT_NAME}_nodes
+  PLUGIN "drivers::serial_driver::SerialBridgeNode"
+  EXECUTABLE "serial_bridge_node_exe"
```
would it be better to name this just `serial_bridge`, since you'll only really see this when calling it by `ros2 run serial_driver`, where `run` signifies an executable?

JWhitleyWork replied:
In-general I just duplicate the class name (`SerialBridgeNode`) and add `_exe` - it's convention in Autoware.Auto too. However, if we do change this, we should change the nodes in `udp_driver` to match.

flynneva replied:
it just seems unnecessarily long. and this name will be what people call via the command line....and usually i appreciate it if they are short 😄 this can be saved for a later time though - not really important ATM

JWhitleyWork replied:
Shortened to just `serial_bridge`. FYI - I expect that most people will use the launch file because it automatically transitions the Lifecycle node but it's not a big deal. Done.

flynneva commented on serial_driver/include/serial_driver/serial_driver.hpp:
```diff
@@ -0,0 +1,46 @@
+public:
+  explicit SerialDriver(const IoContext & ctx);
+
+  void init_port(const std::string & device_name, const SerialPortConfig & config);
```
for the `udp_driver` we have these separated out into a `init_sender` and `init_receiver`. do serial ports by default have both built in? would there be any benefits to separating them out?

JWhitleyWork replied:
The serial protocol, by default, can not connect a port without both send and receive. It's built in all the way down to the wiring (TX and RX wires). Since UDP is, by defnition, a one-way communication mechanism, splitting them makes sense. For serial, I would argue that it does not.

flynneva commented on serial_driver/include/serial_driver/serial_port.hpp:
```diff
@@ -0,0 +1,237 @@
+#include "io_context/common.hpp"
+#include "io_context/io_context.hpp"
+
+using sbp = asio::serial_port_base;
```
should this be `spb`?

JWhitleyWork replied:
It should be, yes. Thanks. Fixed in d524784.

flynneva commented on serial_driver/src/serial_bridge_node.cpp:
```diff
@@ -0,0 +1,218 @@
+
+SerialBridgeNode::SerialBridgeNode(const rclcpp::NodeOptions & options)
+: lc::LifecycleNode("serial_bridge_node", options),
+  m_owned_ctx{new IoContext(2)},
```
does this mean there will be 2 threads always? should this be configurable? asking because honestly I dont know 😄

JWhitleyWork replied:
This means that, if the user doesn't provide their own `IoContext` object with the number of threads pre-defined, the node will create one of its own with 2 threads. It's configurable through the `IoContext` object that the user provides if 2 isn't desired.

flynneva commented on udp_driver/src/udp_sender_node.cpp:
```diff
@@ -141,3 +141,6 @@ void UdpSenderNode::subscriber_callback(std_msgs::msg::Int32::SharedPtr msg)
 }  // namespace udp_driver
 }  // namespace drivers
+
+#include <rclcpp_components/register_node_macro.hpp>  // NOLINT
```
did you mean to include this in this PR? good catch all the same!

JWhitleyWork replied:
I pushed this to your `generic-udp-nodes` branch instead. Thanks.

flynneva commented on serial_driver/package.xml:
```diff
@@ -15,7 +15,10 @@
   <depend>rclcpp</depend>
+  <depend>rclcpp_components</depend>
+  <depend>rclcpp_lifecycle</depend>
   <depend>std_msgs</depend>
```
should there also be `example_interfaces` listed here?

JWhitleyWork replied:
Since this package depends on `io_context` and `io_context` depends on `example_interfaces`, it isn't strictly required. `ament` populates CMake include and library lists recursively.

flynneva commented:
continuing to test this in the mean time to see i can learn where the bug is for my other PR #40.

I added some prints locally to see what is happening. it looks like after setting the lifecycle transition to `configure` the node crashes with a `bad_alloc` error.

first running the node with:
```
ros2 run serial_driver serial_bridge_node_exe --ros-args --params-file /home/ubuntu/code/ros/foxy/src/transport_drivers/serial_driver/params/example.params.yaml 
[INFO] [1616445580.513745094] [IoContext::IoContext]: Thread(s) Created: 2
[INFO] [1616445580.514053849] [serial_bridge_node]: Serial Driver Parameters:
[INFO] [1616445580.514352659] [serial_bridge_node]:   device_name: /dev/ttyUSB0
[INFO] [1616445580.514460268] [serial_bridge_node]:   baud_rate: 115200
[INFO] [1616445580.514556063] [serial_bridge_node]:   flow_control: none
[INFO] [1616445580.514645284] [serial_bridge_node]:   parity: none
[INFO] [1616445580.514737171] [serial_bridge_node]:   stop_bits: 1
```

then from another terminal setting the lifecycle state change with:
```
ros2 lifecycle set /serial_bridge_node configure
Transitioning successful
```

we get this error....i think from the `receive callback` method.

```
terminate called after throwing an instance of 'std::bad_alloc'
  what():  std::bad_alloc
```

JWhitleyWork commented on io_context/include/msg_converters/example_interfaces.hpp:
```diff
@@ -0,0 +1,36 @@
+
+void from_msg(const UInt8MultiArray::SharedPtr & in, MutBuffer & out);
+
+void to_msg(const MutBuffer & in, UInt8MultiArray & out);
```
Yeah, that was my intent. Much more concise and matches the Google Code guidelines.

JWhitleyWork commented:
> ignore my comments above if you already addressed them (i started a review but never submitted it).
> 
> re-compiled and tried to run it and connect to a serial device (a BNO055 IMU) and got this output:
> 
> ```
> ros2 run serial_driver serial_bridge_node_exe --ros-args --params-file ./example.params.yaml 
> [INFO] [1615907120.893310599] [IoContext::IoContext]: Thread(s) Created: 2
> ```
> 
> should we add some prints here to help the user understand at what step the driver is at?
> 
> also it seems like there is only one extra topic for the lifecycle events created. do you think by default for this exe we should start up some sort of async stream & sending so the user could read/write to the serial port via ROS2?

Since this is a LifecycleNode, it has to be transitioned from the start-up state (UNCONFIGURED) to the configured state (CONFIGURED) to the active state (ACTIVE) which is what the launch file does. This is by-design and I don't think we should try to circumvent the design of the Lifecycle nodes. Instead, we should recommend that people use the launch file rather than the node directly if they are either not familiar with Lifecycle nodes or don't want to manage the state of the node directly.

JWhitleyWork commented:
> we get this error....i think from the `receive callback` method.
> 
> ```
> terminate called after throwing an instance of 'std::bad_alloc'
>   what():  std::bad_alloc
> ```

I'm guessing this is in the `to_msg` function. I've made a change to how the memory reservation is done. Would you mind testing it again? If it doesn't work, I'd like to set up a remote session with you, if possible, to see if we can get this fixed real-time since I don't have a good serial device to test with. You can email me at josh.whitley@autoware.org.

reza-ebrahimi commented:
```
terminate called after throwing an instance of 'std::bad_alloc'
  what():  std::bad_alloc
```

I guess the line below causes this exception:
https://github.com/ros-drivers/transport_drivers/blob/serial-driver-updates/serial_driver/include/serial_driver/serial_port.hpp#L230

Change it to a large buffer size and test it.
```cpp
static constexpr size_t m_recv_buffer_size{102400}; // 1024*100
```

JWhitleyWork commented:
> ```
> terminate called after throwing an instance of 'std::bad_alloc'
>   what():  std::bad_alloc
> ```
> 
> I guess the line below causes this exception:
> https://github.com/ros-drivers/transport_drivers/blob/serial-driver-updates/serial_driver/include/serial_driver/serial_port.hpp#L230
> 
> Change it to a large buffer size and test it.
> 
> ```c++
> static constexpr size_t m_recv_buffer_size{102400}; // 1024*100
> ```

This is probably because I did it as a `std::array` instead of a `std::vector`. I've changed this now and also replaced all `MutBuffer` instances with `std::vector<uint8_t>`.

JWhitleyWork commented:
@flynneva, @reza-ebrahimi, and/or @esteve -  Would you mind giving this another look/test?

flynneva commented:
confirmed that it builds OK and runs OK without crashing. i dont have a serial device that just sends out data without being configured first so its a little hard for me to test the receiver function. ill review the code now though

flynneva commented on serial_driver/test/test_serial_port.cpp:
```diff
@@ -0,0 +1,72 @@
+
+#include "serial_driver/serial_port.hpp"
+
+using sbp = asio::serial_port_base;
```
should this be `spb` instead?

JWhitleyWork replied:
AH! Fixed.

flynneva commented on udp_driver/src/udp_socket.cpp:
```diff
@@ -38,6 +38,7 @@ UdpSocket::UdpSocket(
   m_udp_socket(ctx.ios()),
   m_endpoint(address::from_string(ip), port)
 {
+  m_recv_buffer.resize(m_recv_buffer_size);
```
this is a good fix - it was only there for the first packet anyways. good find!

reza-ebrahimi commented:
IMO, This PR #48 should be reverted.

reza-ebrahimi commented:
IMO, Changes related to this issue #48 should be reverted.

flynneva commented:
@reza-ebrahimi do you have an idea how to fix the threading issues we were seeing? please make a PR if you do!

JWhitleyWork commented:
Going ahead and merging. Even though it hasn't been tested on hardware that can talk directly, since no one is actually building based on it yet, we can fix things as we find them.
